File: src/handlers/audit.rs
use axum::{
    extract::{Path, Query, State},
    http::{header, StatusCode},
    response::IntoResponse,
    Extension, Json,
};
use serde::Deserialize;
use std::sync::Arc;
use uuid::Uuid;

use crate::{
    error::{AppError, Result},
    middleware::admin::CurrentUserId,
    models::audit::{
        AlertQuery, AuditLogExportQuery, AuditLogQuery, AuditStats, AuditStatsQuery, ExportFormat,
        UpdateAlertRuleRequest, UpdateAlertStatusRequest,
    },
    models::response::ApiResponse,
    state::AppState,
};

/// 查询审计日志
/// GET /api/v1/admin/audit/logs
pub async fn list_audit_logs(
    State(state): State<Arc<AppState>>,
    Query(query): Query<AuditLogQuery>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let (logs, total) = state.audit_service().query_logs(query.clone()).await?;

    let response = serde_json::json!({
        "logs": logs,
        "total": total,
        "limit": query.limit.unwrap_or(50),
        "offset": query.offset.unwrap_or(0),
    });

    Ok(Json(ApiResponse::success(response)))
}

/// 获取单条审计日志详情
/// GET /api/v1/admin/audit/logs/:id
pub async fn get_audit_log_detail(
    State(state): State<Arc<AppState>>,
    Path(log_id): Path<Uuid>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let log = state
        .audit_service()
        .get_log_by_id(log_id)
        .await?
        .ok_or(AppError::NotFound)?;

    Ok(Json(ApiResponse::success(serde_json::json!(log))))
}

/// 获取审计统计信息
/// GET /api/v1/admin/audit/stats
pub async fn get_audit_stats(
    State(state): State<Arc<AppState>>,
    Query(query): Query<AuditStatsQuery>,
) -> Result<Json<ApiResponse<AuditStats>>> {
    let stats = state
        .audit_service()
        .get_audit_stats(query.start_time, query.end_time)
        .await?;

    Ok(Json(ApiResponse::success(stats)))
}

/// 导出审计日志
/// GET /api/v1/admin/audit/export
pub async fn export_audit_logs(
    State(state): State<Arc<AppState>>,
    Query(query): Query<AuditLogExportQuery>,
) -> Result<impl IntoResponse> {
    let log_query = AuditLogQuery {
        event_type: query.event_type,
        severity: query.severity,
        actor_id: query.actor_id,
        target_id: None,
        target_type: None,
        status: None,
        start_time: query.start_time,
        end_time: query.end_time,
        limit: Some(10000),
        offset: Some(0),
    };

    match query.format {
        ExportFormat::Json => {
            let json_data = state.audit_service().export_logs_json(log_query).await?;
            Ok((
                StatusCode::OK,
                [(header::CONTENT_TYPE, "application/json")],
                [(
                    header::CONTENT_DISPOSITION,
                    "attachment; filename=audit_logs.json",
                )],
                json_data,
            ))
        }
        ExportFormat::Csv => {
            let csv_data = state.audit_service().export_logs_csv(log_query).await?;
            Ok((
                StatusCode::OK,
                [(header::CONTENT_TYPE, "text/csv")],
                [(
                    header::CONTENT_DISPOSITION,
                    "attachment; filename=audit_logs.csv",
                )],
                csv_data,
            ))
        }
    }
}

/// 查询告警列表
/// GET /api/v1/admin/audit/alerts
pub async fn list_alerts(
    State(state): State<Arc<AppState>>,
    Query(query): Query<AlertQuery>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let (alerts, total) = state.audit_service().query_alerts(query.clone()).await?;

    let response = serde_json::json!({
        "alerts": alerts,
        "total": total,
        "limit": query.limit.unwrap_or(50),
        "offset": query.offset.unwrap_or(0),
    });

    Ok(Json(ApiResponse::success(response)))
}

/// 更新告警状态
/// PUT /api/v1/admin/audit/alerts/:id/status
pub async fn update_alert_status(
    State(state): State<Arc<AppState>>,
    Path(alert_id): Path<Uuid>,
    Extension(CurrentUserId(user_id)): Extension<CurrentUserId>,
    Json(request): Json<UpdateAlertStatusRequest>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let alert = state
        .audit_service()
        .update_alert_status(alert_id, request.status, user_id)
        .await?;

    Ok(Json(ApiResponse::success(serde_json::json!(alert))))
}

/// 获取告警规则列表
/// GET /api/v1/admin/audit/rules
pub async fn list_alert_rules(
    State(state): State<Arc<AppState>>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let rules = state.audit_service().get_alert_rules().await?;

    Ok(Json(ApiResponse::success(serde_json::json!({
        "rules": rules
    }))))
}

/// 更新告警规则
/// PUT /api/v1/admin/audit/rules/:id
pub async fn update_alert_rule(
    State(state): State<Arc<AppState>>,
    Path(rule_id): Path<Uuid>,
    Json(request): Json<UpdateAlertRuleRequest>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let rule = state
        .audit_service()
        .update_alert_rule(rule_id, request)
        .await?;

    Ok(Json(ApiResponse::success(serde_json::json!(rule))))
}

/// 清理过期审计日志
/// POST /api/v1/admin/audit/cleanup
#[derive(Debug, Deserialize)]
pub struct CleanupRequest {
    pub days: i64,
    pub archive: Option<bool>,
    pub archive_dir: Option<String>,
}

pub async fn cleanup_audit_logs(
    State(state): State<Arc<AppState>>,
    Json(request): Json<CleanupRequest>,
) -> Result<Json<ApiResponse<serde_json::Value>>> {
    let before = chrono::Utc::now() - chrono::Duration::days(request.days);

    let deleted = if request.archive.unwrap_or(false) {
        // 先归档再删除
        state
            .audit_service()
            .archive_old_logs(before, request.archive_dir.as_deref())
            .await?
    } else {
        // 直接清理
        state.audit_service().cleanup_old_logs(before).await?
    };

    Ok(Json(ApiResponse::success(serde_json::json!({
        "deleted": deleted,
        "before": before,
        "archived": request.archive.unwrap_or(false)
    }))))
}

File: ui-admin/src/api/audit.ts
/**
 * 审计系统 API 模块
 * 提供审计日志、安全告警、告警规则相关接口
 */

import { get, post, put, type ApiResponse } from './client'
import type {
  AuditLog,
  AuditLogQuery,
  AuditLogListResponse,
  AuditStats,
  SecurityAlert,
  AlertQuery,
  AlertListResponse,
  UpdateAlertStatusRequest,
  AlertRule,
  UpdateAlertRuleRequest,
  AlertRuleListResponse,
  AuditLogExportQuery,
} from '@/types'

export type {
  AuditLog,
  AuditLogQuery,
  AuditLogListResponse,
  AuditStats,
  SecurityAlert,
  AlertQuery,
  AlertListResponse,
  UpdateAlertStatusRequest,
  AlertRule,
  UpdateAlertRuleRequest,
  AlertRuleListResponse,
  AuditLogExportQuery,
}

/**
 * 查询审计日志列表
 * @param query 查询参数
 * @returns 日志列表和总数
 */
export async function listAuditLogs(
  query: AuditLogQuery = {}
): Promise<ApiResponse<AuditLogListResponse>> {
  const params = new URLSearchParams()
  if (query.event_type) params.append('event_type', query.event_type)
  if (query.severity) params.append('severity', query.severity)
  if (query.actor_id) params.append('actor_id', query.actor_id)
  if (query.target_id) params.append('target_id', query.target_id)
  if (query.target_type) params.append('target_type', query.target_type)
  if (query.status) params.append('status', query.status)
  if (query.start_time) params.append('start_time', query.start_time)
  if (query.end_time) params.append('end_time', query.end_time)
  if (query.limit) params.append('limit', query.limit.toString())
  if (query.offset) params.append('offset', query.offset.toString())

  const queryString = params.toString()
  return get<AuditLogListResponse>(`/admin/audit/logs${queryString ? `?${queryString}` : ''}`)
}

/**
 * 获取单条审计日志详情
 * @param logId 日志ID
 * @returns 日志详情
 */
export async function getAuditLogDetail(logId: string): Promise<ApiResponse<AuditLog>> {
  return get<AuditLog>(`/admin/audit/logs/${logId}`)
}

/**
 * 获取审计统计信息
 * @param startTime 开始时间
 * @param endTime 结束时间
 * @returns 统计数据
 */
export async function getAuditStats(
  startTime?: string,
  endTime?: string
): Promise<ApiResponse<AuditStats>> {
  const params = new URLSearchParams()
  if (startTime) params.append('start_time', startTime)
  if (endTime) params.append('end_time', endTime)

  const queryString = params.toString()
  return get<AuditStats>(`/admin/audit/stats${queryString ? `?${queryString}` : ''}`)
}

/**
 * 导出审计日志
 * @param query 导出参数
 * @returns Blob 数据
 */
export async function exportAuditLogs(query: AuditLogExportQuery): Promise<Blob> {
  const params = new URLSearchParams()
  params.append('format', query.format)
  if (query.event_type) params.append('event_type', query.event_type)
  if (query.severity) params.append('severity', query.severity)
  if (query.start_time) params.append('start_time', query.start_time)
  if (query.end_time) params.append('end_time', query.end_time)

  const baseUrl = import.meta.env.VITE_API_BASE_URL || '/api/v1'
  const url = `${baseUrl}/admin/audit/export?${params.toString()}`

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${localStorage.getItem('access_token') || sessionStorage.getItem('access_token') || ''}`,
    },
  })

  if (!response.ok) {
    throw new Error('导出失败')
  }

  return response.blob()
}

/**
 * 查询安全告警列表
 * @param query 查询参数
 * @returns 告警列表和总数
 */
export async function listAlerts(
  query: AlertQuery = {}
): Promise<ApiResponse<AlertListResponse>> {
  const params = new URLSearchParams()
  if (query.severity) params.append('severity', query.severity)
  if (query.status) params.append('status', query.status)
  if (query.start_time) params.append('start_time', query.start_time)
  if (query.end_time) params.append('end_time', query.end_time)
  if (query.limit) params.append('limit', query.limit.toString())
  if (query.offset) params.append('offset', query.offset.toString())

  const queryString = params.toString()
  return get<AlertListResponse>(`/admin/audit/alerts${queryString ? `?${queryString}` : ''}`)
}

/**
 * 更新告警状态
 * @param alertId 告警ID
 * @param request 状态更新请求
 * @returns 更新后的告警
 */
export async function updateAlertStatus(
  alertId: string,
  request: UpdateAlertStatusRequest
): Promise<ApiResponse<SecurityAlert>> {
  return put<SecurityAlert>(`/admin/audit/alerts/${alertId}/status`, request)
}

/**
 * 获取告警规则列表
 * @returns 规则列表
 */
export async function listAlertRules(): Promise<ApiResponse<AlertRuleListResponse>> {
  return get<AlertRuleListResponse>('/admin/audit/rules')
}

/**
 * 更新告警规则
 * @param ruleId 规则ID
 * @param request 规则更新请求
 * @returns 更新后的规则
 */
export async function updateAlertRule(
  ruleId: string,
  request: UpdateAlertRuleRequest
): Promise<ApiResponse<AlertRule>> {
  return put<AlertRule>(`/admin/audit/rules/${ruleId}`, request)
}

/**
 * 清理过期审计日志
 * @param days 保留天数
 * @param archive 是否归档
 * @param archiveDir 归档目录
 * @returns 清理的日志数量
 */
export async function cleanupAuditLogs(
  days: number,
  archive?: boolean,
  archiveDir?: string
): Promise<ApiResponse<{ deleted: number }>> {
  return post<{ deleted: number }>('/admin/audit/cleanup', {
    days,
    archive,
    archive_dir: archiveDir,
  })
}
